src/rgw/rgw_lc.cc

```diff
@@ -0,0 +1,546 @@
+#include <string.h>
+#include <iostream>
+#include <map>
+
+#include "include/types.h"
+
+#include "common/Formatter.h"
+#include <common/errno.h>
+#include "auth/Crypto.h"
+#include "include/rados/librados.hpp"
+#include "cls/rgw/cls_rgw_client.h"
+#include "cls/refcount/cls_refcount_client.h"
+#include "cls/lock/cls_lock_client.h"
+#include <common/dout.h>
+#include "rgw_common.h"
+#include "rgw_bucket.h"
+#include "rgw_lc.h"
+#include "rgw_lc_s3.h"
+
+
+
+#define dout_subsys ceph_subsys_rgw
+
+const char* LC_STATUS[] = {
+      "UNINITIAL",
+      "PROCESSING",
+      "FAILED",
+      "COMPLETE"
+};
+
+using namespace std;
+using namespace librados;
+void RGWLifecycleConfiguration::add_rule(LCRule *rule)
+{
+  string id;
+  rule->get_id(id); // not that this will return false for groups, but that's ok, we won't search groups
+  rule_map.insert(pair<string, LCRule>(id, *rule));
+  _add_rule(rule);
+}
+
+void RGWLifecycleConfiguration::_add_rule(LCRule *rule)
+{
+  string prefix;
+  LCExpiration expiration;
+  int days;
+  if (!rule->get_prefix(prefix)) {
+    ldout(cct, 5) << "ERROR: rule->get_prefix() failed" << dendl;
+  }
+  if (!rule->get_expiration(expiration)) {
+    ldout(cct, 5) << "ERROR: rule->get_expiration() failed" << dendl;
+  }
+  if (!expiration.get_days(&days)) {
+    ldout(cct, 5) << "ERROR: expiration->get_days() failed" << dendl;
+  }
+  prefix_map[prefix] = days;
+}
+
+void *RGWLC::LCWorker::entry() {
+  do {
+    utime_t start = ceph_clock_now(cct);
+    if (should_work(start)) {
+      dout(5) << "life cycle: start" << dendl;
+      int r = lc->process();
+      if (r < 0) {
+        dout(0) << "ERROR: do life cycle process() returned error r=" << r << dendl;
+      }
+      dout(5) << "life cycle: stop" << dendl;
+    }
+    if (lc->going_down())
+      break;
+
+    utime_t end = ceph_clock_now(cct);   
+    int secs = shedule_next_start_time(end);
+    time_t next_time = end + secs;
+    char buf[30];
+    char *nt = ctime_r(&next_time, buf);
+    dout(5) << "shedule life cycle next start time: " << nt <<dendl;
+
+    lock.Lock();
+    cond.WaitInterval(cct, lock, utime_t(secs, 0));
+    lock.Unlock();
+  } while (!lc->going_down());
+
+  return NULL;
+}
+
+void RGWLC::initialize(CephContext *_cct, RGWRados *_store) {
```

Comment:
Is there a reason this can't be done in the constructor?  It looks like it should be possible.